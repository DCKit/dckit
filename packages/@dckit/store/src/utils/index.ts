interface AnyObject {
  [key: string]: any
}

// lodash _.get replacement for objects only
export const _get = (obj: AnyObject, path: string, defaultValue?: any): any => {
  if (!obj || !path) return defaultValue

  const parts = String(path).split('.')

  if (parts.length === 1) {
    let value
    try {
      value = obj[parts[0]]
    } catch (e) {
      return defaultValue
    }
    return value
  }

  return parts.reduce((acc, v) => {
    try {
      acc = acc[v]
    } catch (e) {
      return defaultValue
    }
    return acc
  }, obj)
}
