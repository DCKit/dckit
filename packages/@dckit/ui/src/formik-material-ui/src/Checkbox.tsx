import * as React from 'react';
import MuiCheckbox, {
  CheckboxProps as MuiCheckboxProps,
} from '@material-ui/core/Checkbox';
import {
  FieldInputProps,
  FieldMetaProps,
  FieldHelperProps,
  useField,
  useFormikContext,
} from 'formik';

export interface CheckboxProps
  extends Omit<
    MuiCheckboxProps,
    'name' | 'value' | 'error' | 'form' | 'checked' | 'defaultChecked'
  > {
  name: string;
}

export function useFieldToCheckbox<Val = unknown>(
  { disabled, name, ...props }: CheckboxProps,
  customize?: (
    props: [FieldInputProps<Val>, FieldMetaProps<Val>, FieldHelperProps<Val>]
  ) => Partial<Omit<CheckboxProps, 'name'>>
): MuiCheckboxProps {
  const { isSubmitting } = useFormikContext();
  const fieldProps = useField(name);

  const [field] = fieldProps;

  return {
    disabled: disabled ?? isSubmitting,
    ...props,
    ...field,
    // TODO handle indeterminate
    checked: field.value,
    value: field.value ? 'checked' : '',
    ...customize?.(fieldProps),
  };
}

export function Checkbox(props: CheckboxProps) {
  return <MuiCheckbox {...useFieldToCheckbox(props)} />;
}

Checkbox.displayName = 'FormikMaterialUICheckbox';
