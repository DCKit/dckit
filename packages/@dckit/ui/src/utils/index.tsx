import { useMediaQuery } from '@material-ui/core'

export const stub = () => null

export const normalizePath = (path?: string) =>
  !path || path === '/' ? '' : path

export interface IMediaType {
  isDesktop: boolean
  isTablet: boolean
  isMobile: boolean
}

const query = (max: number): string => `@media (max-width: ${max}px)`

export const useMediaType = (): IMediaType => {
  const isDesktop = useMediaQuery(query(922))
  const isTablet = useMediaQuery(query(768))
  const isMobile = useMediaQuery(query(576))
  return {
    isDesktop,
    isTablet,
    isMobile,
  }
}
