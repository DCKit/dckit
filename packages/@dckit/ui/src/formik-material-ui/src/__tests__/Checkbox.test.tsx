import React from 'react';

import { render, fireEvent } from './utils';

import { Checkbox } from '../Checkbox';

test('renders', async () => {
  const onSubmit = jest.fn();
  const { asFragment } = render(<Checkbox name="checked" />, {
    initialValues: {
      checked: false,
    },
    onSubmit,
  });

  expect(asFragment()).toMatchSnapshot();
});

test('checked', async () => {
  // given
  const onSubmit = jest.fn();
  const { getByTestId, findByText } = render(
    <Checkbox
      name="checked"
      label="Checkbox"
      inputProps={{
        // @ts-ignore
        'data-testid': 'checkbox',
      }}
    />,
    {
      onSubmit,
      initialValues: {
        checked: false,
      },
    }
  );

  const input = getByTestId('checkbox');
  const submit = getByTestId('submit');

  // when
  fireEvent.change(input, { target: { value: 'checked' } });
  fireEvent.click(submit);

  await findByText('submitted');

  // then
  expect(onSubmit).toBeCalledWith(
    {
      checked: false,
    },
    expect.anything()
  );
});
