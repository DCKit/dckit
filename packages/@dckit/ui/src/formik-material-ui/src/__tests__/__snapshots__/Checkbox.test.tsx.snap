// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <form
    action="#"
    data-testid="form"
  >
    <span
      aria-disabled="false"
      class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-7 MuiCheckbox-root MuiCheckbox-colorSecondary MuiIconButton-colorSecondary"
    >
      <span
        class="MuiIconButton-label"
      >
        <input
          class="PrivateSwitchBase-input-10"
          data-indeterminate="false"
          name="checked"
          type="checkbox"
          value=""
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </span>
    <button
      data-testid="submit"
    >
      submit
    </button>
  </form>
</DocumentFragment>
`;
